### PR TITLE
[Fix Issue #1197 since 2022] Support pre-trained openai/guided-diffusion (ADM) with minimal code change

### PR DESCRIPTION
**Pull request by tongdaxu**
# What does this PR do?

TLDR: This PR includes openai/guided-diffusion pre-trained models into diffusers, with a diverse family of pre-trained openai/guided-diffusion models, including pre-trained model with ImageNet, LSUN bedroom, LSUN cat, LSUN horse and FFHQ dataset.

The openai/guided-diffusion pre-trained models (ADM unconditional 256x256) are used by academic community a lot (e.g. DPS ICLR 2023: https://github.com/DPS2022/diffusion-posterior-sampling, FreeDOM ICCV 2023: https://github.com/vvictoryuki/FreeDoM?tab=readme-ov-file). However, it is not supported in huggingface/diffusers.

This issue has been raised as early as 2022: https://github.com/huggingface/diffusers/issues/1197 but left unsolved. As it is indeed quite complicated.

I make changes to UNet2DModel as minimal as possible, to make it 
* backward compatiable with previous models
* supports openai/guided-diffusion (ADM)
so that we can play openai/guided-diffusion pretrained model as any standard unconditional diffusion model (https://huggingface.co/docs/diffusers/using-diffusers/unconditional_image_generation).

Those changes includes:
* Interface of UNet2DModel: add new argument "attention_legacy_order"  
* Two necessary building blocks, including:
  * ADM's time_proj, theoretically they are the same as diffusers implementation, but numerically they are different. Replacing one with anothor breaks the model.
  * attention_legacy_order mode in class Attention: the legacy order mode is necessary. use diffusers attention directly breaks the model.

I have been very careful not to break any existing code, and make the new code as short as possible.

I have provided a script to convert pre-trained openai/guided-diffusion to huggingface compatible model, in https://github.com/tongdaxu/diffusers/blob/main/scripts/convert_adm_to_diffusers.py.

I have also provide my conversion of models with configs. Those conversions have mean absolute error $~5e-5$, and relative absolute error $~6e-5$, when the input noise is the same. As the error is minimal, the model and conversion is correct. The complete list of converted models is:
* ImageNet: https://openaipublic.blob.core.windows.net/diffusion/jul-2021/256x256_diffusion_uncond.pt -> https://huggingface.co/xutongda/adm_imagenet_256x256_unconditional
* LSUN bedroom: https://openaipublic.blob.core.windows.net/diffusion/jul-2021/lsun_bedroom.pt -> https://huggingface.co/xutongda/adm_lsun_bedroom_256x256
* LSUN cat: https://openaipublic.blob.core.windows.net/diffusion/jul-2021/lsun_cat.pt -> https://huggingface.co/xutongda/adm_lsun_cat_256x256
* LSUN horse: https://openaipublic.blob.core.windows.net/diffusion/jul-2021/lsun_horse.pt -> https://huggingface.co/xutongda/adm_lsun_horse_256x256
* FFHQ: https://drive.google.com/drive/folders/1jElnRoFv7b31fG0v6pTSQkelbSX3xGZh -> https://huggingface.co/xutongda/adm_ffhq_256x256/tree/main

Now we can sample from the pre-trained models of openai/guided-diffusion, using diffusers in an out of box way.

```python
from diffusers import DiffusionPipeline

generator = DiffusionPipeline.from_pretrained("xutongda/adm_imagenet_256x256_unconditional").to("cuda")
image = generator().images[0]
image.save("generated_image.png")
```

And the result is as good as the original openai/guided-diffusion model:
* sample in diffusers with converted model
sample
* sample in openai/guided-diffusion with original model
sample_adm

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @yiyixuxu and @sayakpaul

### CODE REVIEW

**Comment by tongdaxu** [comment]
Some other samples using the converted model with diffusers:
* Samples from LSUN cat model
generated_image_cat
* Samples from FFHQ model
generated_image_ffhq
* Samples from LSUN horse model
generated_image_horse
* Samples from LSUN bedroom model
generated_image_bedroom

**Comment by sayakpaul** [comment]
Thanks very much for your work on this. 

I agree that ADM is still very much used by the academic community but probably doesn't have a lot of real-world significance because of the lower quality. On the other hand, we do support Consistency Models as well as the original DDPM and DDIM models to respect the literature. 

So, given the above point and also considering the minimal changes introduced in this PR, I'd be supportive of adding it. My only major feedback would be to try to not use legacy attention blocks if possible. 

@patrickvonplaten @yiyixuxu WDYT here?

**Comment by tongdaxu** [comment]
> Thanks very much for your work on this.
> 
> I agree that ADM is still very much used by the academic community but probably doesn't have a lot of real-world significance because of the lower quality. On the other hand, we do support Consistency Models as well as the original DDPM and DDIM models to respect the literature.
> 
> So, given the above point and also considering the minimal changes introduced in this PR, I'd be supportive of adding it. My only major feedback would be to try to not use legacy attention blocks if possible.
> 
> @patrickvonplaten @yiyixuxu WDYT here?

The problem here is that all the offical pre-trained ADM by openai use legacy attention, so I really have no choice but using them. I have tried to use diffuser attention but the model produces garbage images like (suppose to be bedroom):

generated_image_bedroom_bad

**Comment by sayakpaul** [comment]
> The problem here is that all the offical pre-trained ADM by openai use legacy attention, so I really have no choice but using them.

Can we try to maybe find a way to port the legacy attention to the one that's used now?

**Comment by tongdaxu** [comment]
> > The problem here is that all the offical pre-trained ADM by openai use legacy attention, so I really have no choice but using them.
> 
> Can we try to maybe find a way to port the legacy attention to the one that's used now?

Sorry, I did not quite get what you mean by "port". Did you mean to create a separate class like legacy attention, and use arguement like attention_type?

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -1234,11 +1239,16 @@ def __call__(
...
-
-        key = key.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
-        value = value.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
-
+        if attn.attention_legacy_order:
+            qkv = torch.cat([query, key, value], dim=2).transpose(1, 2)
+            query, key, value = qkv.reshape(batch_size, attn.heads, head_dim * 3, -1).chunk(3, dim=2)
+            query = query.transpose(-1, -2)
+            key = key.transpose(-1, -2)
+            value = value.transpose(-1, -2)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:1247]: We should be able to get rid of this if we incorporate this during the conversion. See how DiT does it:

https://github.com/huggingface/diffusers/blob/acd1962769715cebee1bb8e87075db4e78490070/scripts/convert_dit_to_diffusers.py#L58

**Comment by sayakpaul** [comment]
See my comment here https://github.com/huggingface/diffusers/pull/6730/files#r1468858192

**Comment by tongdaxu** [comment]
> See my comment here https://github.com/huggingface/diffusers/pull/6730/files#r1468858192

In fact, the part you refer to is about model conversion only, and I have already done it by calling the code of https://github.com/tongdaxu/diffusers/blob/main/scripts/convert_consistency_to_diffusers.py#L143. In this way, we can indeed unify the model weights of openai/guideddiffusion and diffusers. However, it has nothing to do with legacy / non legacy attention order. It is purely about the parameterization of linear layers.

However, what can not be avoided is the run time difference between legacy and non legacy attention. The "qkv, q, k, v" you are referring to are model weights, the "qkv, q, k, v" I am referring to are activation tensors. They are different stuffs with different shape.

**Comment by tongdaxu** [comment]
> See my comment here https://github.com/huggingface/diffusers/pull/6730/files#r1468858192

In openai/guideddiffusion, both normal attention and legacy attention are implemented in separate class:
* normal: https://github.com/openai/guided-diffusion/blob/main/guided_diffusion/unet.py#L361
* legacy: https://github.com/openai/guided-diffusion/blob/main/guided_diffusion/unet.py#L328

Those two attentions have exactly the same model weights, and the two classes have no parameters at all. So it has nothing to do with clever tricks in model conversion. This has to be solved in runtime.

**Comment by tongdaxu** [comment]
Hi @sayakpaul, any further comments?

File: src/diffusers/models/unets/unet_2d_blocks.py [hunk 2]
```diff
@@ -549,21 +554,22 @@ def __init__(
...
+        attention_legacy_order: bool = False,
         output_scale_factor: float = 1.0,
     ):
         super().__init__()
         resnet_groups = resnet_groups if resnet_groups is not None else min(in_channels // 4, 32)
         self.add_attention = add_attention
 
         if attn_groups is None:
-            attn_groups = resnet_groups if resnet_time_scale_shift == "default" else None
+            attn_groups = None if resnet_time_scale_shift == "spatial" else resnet_groups
```
Comment (yiyixuxu) [inline src/diffusers/models/unets/unet_2d_blocks.py:572]: this will break for `resnet_time_scale_shift = "scale_shift"`, no? Even though the doc string says only accept `"default"` and `"spatial"` I'm not sure if no model has configured with `resnet_time_scale_shift = scale_shift`

I want to know your thoughts here @sayakpaul

Comment (sayakpaul) [reply src/diffusers/models/unets/unet_2d_blocks.py:572]: Can we maybe come back with a better condition that doesn't leave any grounds for a potentially backward-breaking change? I agree with @yiyixuxu here.

Comment (tongdaxu) [reply src/diffusers/models/unets/unet_2d_blocks.py:572]: > Can we maybe come back with a better condition that doesn't leave any grounds for a potentially backward-breaking change? I agree with @yiyixuxu here.

Cool, let me see if I can solve it in some other way. Would you prefer the current solution, or a slightly more ugly one like a new class of UNetMiddleBlock? In fact, I can easy make it free of backward-breaking, but I might introduce more ugly additional blocks with more code.

**Comment by tongdaxu** [comment]
Hi, I have found a way to avoid breaking the possible backward deps in class UNetMidBlock2D and updated the PR. The change is still minimal but it does not break anything.

I would love your advices @yiyixuxu @sayakpaul.

File: src/diffusers/models/embeddings.py [hunk 3]
```diff
@@ -254,6 +254,38 @@ def forward(self, timesteps):
...
+
+    def __init__(self, num_channels: int):
+        super().__init__()
+        self.num_channels = num_channels
+
+    def forward(self, timesteps):
+        t_emb = timestep_embedding_adm(
+            timesteps,
+            self.num_channels,
+        )
```
Comment (sayakpaul) [inline src/diffusers/models/embeddings.py:285]:
```suggestion
        t_emb = timestep_embedding_adm(timesteps, self.num_channels)
```

File: scripts/convert_adm_to_diffusers.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    ],
+    "up_block_types": [
+        "ResnetUpsampleBlock2D",
+        "AttnUpBlock2D",
+        "ResnetUpsampleBlock2D",
+        "ResnetUpsampleBlock2D",
+        "ResnetUpsampleBlock2D",
+        "ResnetUpsampleBlock2D",
+    ],
+    "resnet_time_scale_shift": "scale_shift",
```
Comment (sayakpaul) [inline scripts/convert_adm_to_diffusers.py:36]: Could you show me the codepath that becomes effective when `resnet_time_scale_shift == "scale_shift"`?

**Review by sayakpaul — COMMENTED** [review]
Looking really lean. I left a question.

File: scripts/convert_adm_to_diffusers.py (hunk 4 above)
Comment (tongdaxu) [reply scripts/convert_adm_to_diffusers.py:36]: The convert_adm_to_diffusers.py is just calling functions of con_pt_to_diffuser function in convert_consistency_to_diffusers.py.

First, the resnet_time_scale_shift == "scale_shift" is not a new option set by this PR. 

Setting resnet_time_scale_shift == "scale_shift" will pass the argument though UNet2DModel. Through init of UNet2DModel, it will be pass into get_down_block, UNetMidBlock2D and get_up_block, and will be further passed into each building blocks such as ResnetDownsampleBlock2D, ResnetUpsampleBlock2D, AttnDownBlock2D, AttnUpBlock2D. The eventual effect of resnet_time_scale_shift == "scale_shift" will set the class ResnetBlock2D's time_embedding_norm == "scale_shift". And this option effects the resnet's time embedding's shape https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/resnet.py#L283, and the behaviour of time embedding https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/resnet.py#L378. 

It has no special effect on UNetMidBlock2D, as I circumvent this problem in https://github.com/tongdaxu/diffusers/blob/main/src/diffusers/models/unets/unet_2d.py#L200.

The resnet_time_scale_shift == "scale_shift" is necessary in model conversion script as the resnet's time embedding's input shape is doubled with resnet_time_scale_shift == "scale_shift".

Comment (sayakpaul) [reply scripts/convert_adm_to_diffusers.py:36]: That makes sense, thanks!

**Comment by patrickvonplaten** [comment]
Let's add this model to the research_projects folder no? It's a bit too outdated to be in core diffusers I'd say (cc @yiyixuxu)

**Comment by yiyixuxu** [comment]
@tongdaxu 
can we move this to the research folder?

**Comment by sayakpaul** [comment]
> I am fine with that, what should I do to move this to research folder?

You can follow the structure of https://github.com/huggingface/diffusers/tree/main/examples/research_projects/controlnetxs as an example. Here's what you could consider.

Have all the conversion script, modeling and pipeline files under a folder and make sure they work.

**Comment by kschwethelm** [comment]
Hi @tongdaxu, thank you for your great work! 

I have issues generating good samples, I hope you can help.

I installed your PR as follows:

1. `conda create -n newenv python=3.9`
2. `conda activate newenv`
3. `pip install torch==1.11.0+cu113 torchvision==0.12.0+cu113 torchaudio==0.11.0 --extra-index-url https://download.pytorch.org/whl/cu113`
4. `pip install git+https://github.com/huggingface/diffusers.git@refs/pull/6730/head`

Then I ran:

```python
from diffusers import DiffusionPipeline

generator = DiffusionPipeline.from_pretrained("xutongda/adm_ffhq_256x256").to("cuda")
image = generator().images[0]
image.save("generated_image.png")
```

With this I get strange generations, e.g.:

generated_image

Did I install the PR wrong or is there a bug?

**Comment by kschwethelm** [comment]
Hi @tongdaxu, thank you for your great work!

I am having trouble generating nice images with your PR. I hope you can help :)

I installed the PR as follows:
1. `conda create -n newenv python=3.9`
2. `conda activate newenv`
3. `pip install torch==1.11.0+cu113 torchvision==0.12.0+cu113 torchaudio==0.11.0 --extra-index-url https://download.pytorch.org/whl/cu113`
4. `pip install git+https://github.com/huggingface/diffusers.git@refs/pull/6730/head`

Then i ran:
```python
from diffusers import DiffusionPipeline

generator = DiffusionPipeline.from_pretrained("xutongda/adm_imagenet_256x256_unconditional").to("cuda")
image = generator().images[0]
image.save("generated_image.png")
```

My generated images look like this:

generated_image

Did I install the PR wrong or is there a bug?

**Comment by tongdaxu** [comment]
> Hi @tongdaxu, thank you for your great work!
> 
> I am having trouble generating nice images with your PR. I hope you can help :)
> 
> I installed the PR as follows:
> 
> 1. `conda create -n newenv python=3.9`
> 2. `conda activate newenv`
> 3. `pip install torch==1.11.0+cu113 torchvision==0.12.0+cu113 torchaudio==0.11.0 --extra-index-url https://download.pytorch.org/whl/cu113`
> 4. `pip install git+https://github.com/huggingface/diffusers.git@refs/pull/6730/head`
> 
> Then i ran:
> 
> ```python
> from diffusers import DiffusionPipeline
> 
> generator = DiffusionPipeline.from_pretrained("xutongda/adm_imagenet_256x256_unconditional").to("cuda")
> image = generator().images[0]
> image.save("generated_image.png")
> ```
> 
> My generated images look like this:
> 
> generated_image
> 
> Did I install the PR wrong or is there a bug?

I am out of office until 17 Feb, would you like to try other models first? And the hugging face model hub has been updated since the middle of commit. Are you using the latest model?

**Comment by tongdaxu** [comment]
> Hi @tongdaxu, thank you for your great work!
> 
> I am having trouble generating nice images with your PR. I hope you can help :)
> 
> I installed the PR as follows:
> 
> 1. `conda create -n newenv python=3.9`
> 2. `conda activate newenv`
> 3. `pip install torch==1.11.0+cu113 torchvision==0.12.0+cu113 torchaudio==0.11.0 --extra-index-url https://download.pytorch.org/whl/cu113`
> 4. `pip install git+https://github.com/huggingface/diffusers.git@refs/pull/6730/head`
> 
> Then i ran:
> 
> ```python
> from diffusers import DiffusionPipeline
> 
> generator = DiffusionPipeline.from_pretrained("xutongda/adm_imagenet_256x256_unconditional").to("cuda")
> image = generator().images[0]
> image.save("generated_image.png")
> ```
> 
> My generated images look like this:
> 
> generated_image
> 
> Did I install the PR wrong or is there a bug?

Sorry I do not have access to GPU for now. But the instructions in https://github.com/tongdaxu/InverseDiffusion should work. Would you like to give it a try?

**Comment by kschwethelm** [comment]
Thank you for your quick response. Sadly, your instructions did not work either.

I tried all versions of your repository and different pretrained models, but I still get bad results.

**Comment by tongdaxu** [comment]
> Thank you for your quick response. Sadly, your instructions did not work either.
> 
> I tried all versions of your repository and different pretrained models, but I still get bad results, e.g., FFHQ:
> 
> generated_image

I believe it is a bug on my side. It could be the last force push. I will fix it ASAP.

**Comment by tongdaxu** [comment]
> Thank you for your quick response. Sadly, your instructions did not work either.
> 
> I tried all versions of your repository and different pretrained models, but I still get bad results, e.g., FFHQ:
> 
> generated_image

Hi, I just ran a small test with https://github.com/tongdaxu/diffusers/commit/111eac139f3dc6ff47c50810a35044cad9b323b1. It seems to be ok.

捕获

I am not sure what is happening here. And I might need more time figuring it out when I am back to office after 17 Feb.

I can only run some small sanity check for now. And all I can say is that with the commit above and the imagenet model (I check the hash sum), the sampling should be fine. I am not sure if there can be some dependency problem (I am using torch 2.1.0). I might need to go back to office for more testing.

Thanks for pointing it out and for your patience.

**Comment by tongdaxu** [comment]
> 4\. pip install git+https://github.com/huggingface/diffusers.git@refs/pull/6730/head

Thank you @kschwethelm, that is very strange. I also find that it fails with torch 1.9 and works with torch 2.1. 

I do not have a clue about why it fails. Have you figured out why?

**Comment by tongdaxu** [comment]
> Hi, thank you very much for your help! The problem was actually the PyTorch version. I now tried with torch-2.2.0 and it works fine.
> 
> generated_image

I can't remember I have added any torch version sensitive code.

**Comment by sayakpaul** [comment]
Not stale. @yiyixuxu WDYT?

**Comment by jiangyuhangcn** [comment]
how to finetune with the adm pretrained model

**Comment by EdwardLeeMacau** [comment]
Hi, thanks for raising this issue. I’ve also encountered the challenge when attempting to migrate the model to diffusers. 

For the attention operations, I found that a possible solution is to reorder the weights in ADM’s `AttentionBlock.qkv`, it avoids to add a legacy model. Here’s an example for converting ImageNet1K model:

```python
def convert_attention(checkpoint, new_checkpoint, old_prefix, new_prefix, attention_dim=None):
    def reorder(t, offset):
        """
        In the legacy attention format, Q, K, and V are stored as:
            [Q1, K1, V1, Q2, K2, V2, ..., Qn, Kn, Vn] 
        with each slice of size 64.

        To migrate to the newer format, we rearrange this as:
            [Q1, Q2, ..., Qn, K1, K2, ..., Kn, V1, V2, ..., Vn]

        This allows us to merge the weights of m2 into m1:
            [Q, K, V]
        """
        return torch.cat([t[3*i+offset:3*i+offset+64] for i in range(0, 1024, 64)]).squeeze(-1)

    weight_q = reorder(checkpoint[f"{old_prefix}.qkv.weight"], 0)
    weight_k = reorder(checkpoint[f"{old_prefix}.qkv.weight"], 64)
    weight_v = reorder(checkpoint[f"{old_prefix}.qkv.weight"], 128)
    bias_q = reorder(checkpoint[f"{old_prefix}.qkv.bias"], 0)
    bias_k = reorder(checkpoint[f"{old_prefix}.qkv.bias"], 64)
    bias_v = reorder(checkpoint[f"{old_prefix}.qkv.bias"], 128)

    # Continue with the remaining steps ...
```

The magic number `64` refers to the number of channels per head, and `1024` means total number of channels.

Some minor revision in my local env:
```json
{
     "attn_norm_num_groups": 32,
     "norm_eps": 1e-05,
}
```

I have checked that the model has mean absolute error $8.5e-5$ (idk why its larger than @tongdaxu provides one). The following images are generated using DDPMScheduler. 
generated_image

If this solution looks good to you, I’d be happy to help finish the remaining work in this PR. However, I may need to wait until around late November to complete it.

**Comment by dapaoA** [comment]
@tongdaxu Hi! I am trying to do inverse problem on ADMs and am glad to be lucky to find your branch, I try to use your code and rebased it to the current diffusers. However, I got the same problem as @kschwethelm. I tried torch 2.5.1+cu121. and also torch2.1.0+cu118 on windows. do you have any new ideas on the reason of this problem?

**Comment by Bili-Sakura** [comment]
@sayakpaul @yiyixuxu request for this feature as a community pipeline integrated and along with diffusers style checkpoint;
I am do with help @tongdaxu
